source/base/incremental_function.cc

```diff
@@ -0,0 +1,101 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2007 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/incremental_function.h>
+
+#include <deal.II/lac/vector.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Functions
+{
+  template <int dim, typename RangeNumberType>
+  IncrementalFunction<dim, RangeNumberType>::IncrementalFunction(
+    Function<dim, RangeNumberType> &base)
+    : base(base)
+    , delta_t(time_type())
+    , values_old(base.n_components)
+  {}
+
+
+
+  template <int dim, typename RangeNumberType>
+  void
+  IncrementalFunction<dim, RangeNumberType>::set_decrement(
+    const time_type delta)
+  {
+    Assert(delta >= 0.0,
+           ExcMessage("The decrement must be set to a non-negative value."));
+    delta_t = delta;
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  RangeNumberType
+  IncrementalFunction<dim, RangeNumberType>::value(const Point<dim> &p,
+                                                   unsigned int      comp) const
+  {
+    // Cache the time state of the base class in case it has been changed
+    // within the user code. We reset the wrapped function to the original
+    // state once we're done with our own evaluations.
+    const auto orig_time = base.get_time();
+
+    base.set_time(this->get_time());
+    const RangeNumberType current = base.value(p, comp);
+
+    base.set_time(this->get_time() - delta_t);
+    const RangeNumberType old = base.value(p, comp);
+
+    // Reset wrapped function time setting
+    base.set_time(orig_time);
+
+    return current - old;
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  void
+  IncrementalFunction<dim, RangeNumberType>::vector_value(
+    const Point<dim> &       p,
+    Vector<RangeNumberType> &values) const
+  {
+    // since we modify a mutable member variable, lock the
+    // the data via a mutex
+    std::lock_guard<std::mutex> lock(mutex);
+
+    // Cache the time state of the base class in case it has been changed
+    // within the user code. We reset the wrapped function to the original
+    // state once we're done with our own evaluations.
+    const auto orig_time = base.get_time();
+
+    base.set_time(this->get_time());
+    base.vector_value(p, values);
+
+    base.set_time(this->get_time() - delta_t);
+    base.vector_value(p, values_old);
```

Comment:
As a side note (aka "opinion piece"):

I am not a huge fan of the original design of `Function` with a "global" time state I have to say. The fact that this function requires a mutex in all its `value` functions is a severe issue: We are advertising thread parallelism (and we nowadays have small compute servers with north of 40 cores or so) and this function will essentially force every tight loop that calls it to run in serial.

It might be time to rethink the whole `dealii::Function` business anyway. I nowadays use small wrappers around `std::function` and lambdas in my own code (that are simply wrapped into `dealii::Function` objects) (I will try to make a WIP pull request for those later today to start a discussion about it).

Maybe given the fact that @davydden needs such an incremental function (and I absolutely agree that we should have an easy way of expressing this concept in deal.II) it is finally time to clean up time handling in our function objects (aka make it not a global state). It might be possible to go the `LinearOperator` route here (that managed to replace a lot of single purpose classes), @luca-heltai thoughts?

Reply:
Example: Suppose you have a function
```
std::function<double(double /*time*/, Point<dim> /*point*/) foo;
```
You can make this an incremental function by simply wrapping in a lambda:
```
double t, delta_t; /* assume these are class members, or local variables */
auto incremental_foo = [&t, &delta_t](Point<dim> point) {
  return foo(t + delta_t, point) - foo(t, point);
  };
```
The only piece that is missing in the library would be a wrapper `to_function` that creates a deal.II function out of that:
```
auto wrapper_incremental_function = to_function(incremental_foo);
```

How does that look?

Reply:
> How does that look?

that looks nice, but my usage is different. I want to use `FunctionParser` to read user-provided Dirichlet BC and then use them in the code in incremental form. So I need to wrap `deal.II` function, not the `std::function`.

But I do agree with your comment above about the "global" time state.

Reply:
There is already a class that takes a regular function and makes a `Function` object out of it. It's in `namespace Functions` (but I forgot the name).

Reply:
As for the mutex, you will have noticed that @davydden had meant the mutex originally just to guard the temp vector -- but this of course doesn't make the situation any better.

I agree in general, but am not overly worried about this particular case here.

Reply:
@bangerth You mean two special case wrappers `ScalarFunctionFromFunctionObject`, and `VectorFunctionFromFunctionObject`, and `VectorFunctionFromTensorFunction`.

Reply:
Yes.